tsl/src/backup.c

```diff
@@ -0,0 +1,177 @@
+{
+	Datum values[_Anum_restore_point_max];
+	bool nulls[_Anum_restore_point_max] = { false };
+	HeapTuple tuple;
+
+	tupdesc = BlessTupleDesc(tupdesc);
+	values[AttrNumberGetAttrOffset(Anum_restore_point_node_name)] = CStringGetDatum(node_name);
+	if (node_name == NULL)
+		nulls[AttrNumberGetAttrOffset(Anum_restore_point_node_name)] = true;
+	values[AttrNumberGetAttrOffset(Anum_restore_point_lsn)] = LSNGetDatum(lsn);
+	tuple = heap_form_tuple(tupdesc, values, nulls);
+	return HeapTupleGetDatum(tuple);
+}
+
+Datum
+create_distributed_restore_point(PG_FUNCTION_ARGS)
+{
+	const char *name = TextDatumGetCString(PG_GETARG_DATUM(0));
+	DistCmdResult *result_cmd;
+	FuncCallContext *funcctx;
+	XLogRecPtr lsn;
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		MemoryContext oldctx;
+		TupleDesc tupdesc;
+		char *sql;
+
+		if (strlen(name) >= MAXFNAMELEN)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("restore point name is too long"),
+					 errhint("maximum %d characters", MAXFNAMELEN - 1)));
+
+		if (RecoveryInProgress())
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 (errmsg("recovery is in progress"),
+					  errhint("WAL control functions cannot be executed during recovery."))));
+
+		if (!XLogIsNeeded())
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 errmsg("WAL level is not sufficient for creating a restore point"),
+					 errhint(
+						 "wal_level must be set to \"replica\" or \"logical\" at server start.")));
+
+		if (!superuser())
+			ereport(ERROR,
+					(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					 errmsg("must be superuser to create restore point")));
+
+		if (!ts_guc_enable_2pc)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_OPERATION_NOT_SUPPORTED),
+					 errmsg("two-phase commit transactions are not enabled"),
+					 errhint("timescale.enable_2pc must be set to true.")));
+
+		if (dist_util_membership() != DIST_MEMBER_ACCESS_NODE)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_OPERATION_NOT_SUPPORTED),
+					 errmsg("distributed restore point must be created on the access node"),
+					 errhint("pg_create_restore_point() can be used to to create restore point "
+							 "on the data node.")));
+
+		/*
+		 * In order to achieve synchronization across the multinode cluster,
+		 * we must ensure that the restore point created on the access node is
+		 * synchronized with each data node.
+		 *
+		 * We must avoid possible situation when a concurrent remote transaction
+		 * commits on the access node during the creation of the distributed
+		 * restore point and gets into the data nodes WAL before the
+		 * restore point marker written.
+		 *
+		 * To do that we take an exclusive lock on the remote transaction
+		 * table, which will force any concurrent transaction
+		 * wait during their PREPARE phase.
+		 */
+		LockRelationOid(ts_catalog_get()->tables[REMOTE_TXN].id, ExclusiveLock);
```

Comment:
I guess this lock level is OK, but, FWIW, the minimum lock level to conflict with other distributed transactions is `RowExclusiveLock`. Do we really need to take anything heavier than that? (i.e., this blocks things like VACUUM, etc.)

Reply:
I guess it is fine to downgrade it as long as it does what it is intended to, I'll give it a try

Reply:
Correction: the minimum lock level to conflict I believe would be `ShareRowLock`. So, although `ExclusiveLock` certainly works, it is strictly higher than necessary. Perhaps this doesn't matter for this case, though, since currently we know the processes involved. Anyway, do what you deem best here.